lib/seadsa/SeaDsaAliasAnalysis.cc

```diff
@@ -1,41 +1,207 @@
+      DOG(llvm::errs()
+	  << "SeaDsaAA: cannot handle functions in different modules.\n");
+      return nullptr;
+    }
+  }
+  return M;
+}
+
+static uint64_t storageSize(const Type *t, const DataLayout &dl) {
+  return dl.getTypeStoreSize(const_cast<Type *>(t));
+}
+
+static Optional<uint64_t> sizeOf(const Graph::Set &types, const DataLayout &dl) {
+  if (types.isEmpty()) {
+    return 0;
+  } else {
+    uint64_t sz = storageSize(*(types.begin()), dl);
+    if (types.isSingleton()) {
+      return sz;
+    } else {
+      auto it = types.begin();
+      ++it;
+      if (std::all_of(it, types.end(), [dl, sz](const Type *t) {
+	      return (storageSize(t, dl) == sz);
+          })) {
+        return sz;
+      } else {
+        return None;
+      }
+    }
+  }
+}
+  
+static bool mayAlias(const Cell &c1, const Cell &c2, const DataLayout &dl) {
+  auto maybeUnsafe = [](const Node *n) {
+		       return n->isIntToPtr() || n->isPtrToInt() ||
+			 n->isIncomplete() || n->isUnknown();
+		     };
+  
+  if (c1.isNull() || c2.isNull()) {
+    return true;
+  }
+  
+  const Node *n1 = c1.getNode();
+  const Node *n2 = c2.getNode();
+
+  if (maybeUnsafe(n1) || maybeUnsafe(n2)) {
+    return true;
+  }
+
+  if (n1 != n2) {
+    // different nodes cannot alias
+    return false;
+  }
+
+  if (n1->isOffsetCollapsed() || n1->isArray()) {
+    return true;
+  }
+
+  unsigned o1, o2;
+  
+  if (c1.getOffset() <= c2.getOffset()) {
+    o1 = c1.getOffset();
+    o2 = c2.getOffset();
+  } else {
+    o1 = c2.getOffset();
+    o2 = c1.getOffset();
+  } 
+
+  assert(o1 <= o2);
+  
+  if (!n1->hasAccessedType(o1)) {
+    return true;
+  }
+
+  auto sizeOfOffset1 = sizeOf(n1->getAccessedType(o1), dl);
+  if (!sizeOfOffset1.hasValue()) {
+    return true;
+  }
+  
+  // if offsets can overlap then may alias
+  return (o1 + sizeOfOffset1.getValue()) >= o2;
+}
+
+			
 llvm::AliasResult SeaDsaAAResult::alias(const llvm::MemoryLocation &LocA,
                                         const llvm::MemoryLocation &LocB, llvm::AAQueryInfo &AAQI) {
-  DOG(llvm::errs() << "\n\n\nAlias query: " << *LocA.Ptr << " and " << *LocB.Ptr << "\n\n\n";);
-  if (LocA.Ptr == LocB.Ptr) return MustAlias;
-  // -- fall back to default implementation
+  DOG(llvm::errs() << "SeaDsaAA --- Alias query: " << *LocA.Ptr << " and " << *LocB.Ptr << "\n\n";);
+
+  auto *ValA = const_cast<Value *>(LocA.Ptr);
+  auto *ValB = const_cast<Value *>(LocB.Ptr);
+
+  if (!ValA->getType()->isPointerTy() || !ValB->getType()->isPointerTy()) {
+    return NoAlias;
+  }
+  
+  if (ValA == ValB) {
+    return MustAlias;
+  }
+
+  const DataLayout *dl = nullptr;
+  // Run seadsa if we have not done it yet
+  if (!m_dsa) {
+    if (Module *M = getModuleFromQuery(ValA, ValB)) {
+      m_fac = std::make_unique<Graph::SetFactory>();
+      dl = &(M->getDataLayout());
+      CallGraph cg(*M); 
+      m_dsa = std::make_unique<BottomUpTopDownGlobalAnalysis>(*dl, m_tliWrapper, m_awi, cg, *m_fac);
```

Comment:
`cg` is deallocated when this method exits. Is it really not stored by `GlobalAnalysis`? I would keep `cg` in a class field, just in case.